tools/crm_shadow.c

```diff
@@ -93,453 +145,568 @@ shadow_teardown(char *name)
-        "\t(Advanced) Switch to the named shadow copy", pcmk__option_default
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\nAdditional Options:", pcmk__option_default
-    },
-    {
-        "force", no_argument, NULL, 'f',
-        "\t\t(Advanced) Force the action to be performed", pcmk__option_default
-    },
-    {
-        "batch", no_argument, NULL, 'b',
-        "\t\t(Advanced) Don't spawn a new shell", pcmk__option_default
-    },
-    {
-        "all", no_argument, NULL, 'a',
-        "\t\t(Advanced) Upload entire CIB, including status, with --commit",
-        pcmk__option_default
-    },
-    {
-        "validate-with", required_argument, NULL, 'v',
-        "(Advanced) Create an older configuration version", pcmk__option_default
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "\nExamples:", pcmk__option_paragraph
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "Create a blank shadow configuration:", pcmk__option_paragraph
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        " crm_shadow --create-empty myShadow", pcmk__option_example
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "Create a shadow configuration from the running cluster:",
-        pcmk__option_paragraph
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        " crm_shadow --create myShadow", pcmk__option_example
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "Display the current shadow configuration:", pcmk__option_paragraph
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        " crm_shadow --display", pcmk__option_example
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "Discard the current shadow configuration (named myShadow):",
-        pcmk__option_paragraph
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        " crm_shadow --delete myShadow --force", pcmk__option_example
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        "Upload current shadow configuration (named myShadow) "
-            "to running cluster:",
-        pcmk__option_paragraph
-    },
-    {
-        "-spacer-", no_argument, NULL, '-',
-        " crm_shadow --commit myShadow", pcmk__option_example
-    },
-    { 0, 0, 0, 0 }
+static bool
+cmd_is_dangerous(enum shadow_command cmd)
+{
+    switch (cmd) {
+        case shadow_cmd_commit:
+        case shadow_cmd_delete:
+            return true;
+        default:
+            return false;
+    }
+}
+
+static gboolean
+set_shadow_from_env(const gchar *option_name, GError **error)
+{
+    pcmk__str_update(&options.shadow, getenv("CIB_shadow"));
+    if (options.shadow == NULL) {
+        exit_code = CRM_EX_NOSUCH;
+        g_set_error(error, PCMK__EXITC_ERROR, exit_code,
+                    "No active shadow configuration defined for command '%s'",
+                    option_name);
+        return FALSE;
+    }
```

Comment:
To elaborate a bit on how error vs last wins intersects option processing vs real work, it makes more sense with last-wins to do this sort of validation after processing all options, since the option that gets us here might be overridden by a later option.

Reply:
That makes sense. And obviously I forgot to come back and update this commit after your feedback on parsing-vs-work in the other PR.

Reply:
By the way, we're prone to that same issue for a lot of the commands regardless. Most of them (a little over half) take the shadow name from their optarg rather than from the environment, and that requires a string copy.

Reply:
About to push an update... the only real downside of moving this out of the option parsing is that we can no longer be as specific and log `"for command '%s'", option_name`. Unless we add an enum2string mapping, which is easy but may not be worth it just for the log message.